test/com/amazon/ion/impl/IonReaderLookaheadBufferTestBase.java

```diff
@@ -0,0 +1,675 @@
+                        assertEquals(IonType.INT, reader.next());
+                        assertEquals(123, reader.intValue());
+                        assertNull(reader.next());
+                        reader.stepOut();
+                    }
+                }
+            ),
+            new Value(
+                toBytes("()"),
+                new IonReaderAssertion() {
+                    @Override
+                    public void evaluate(IonReader reader) {
+                        assertEquals(IonType.SEXP, reader.next());
+                        reader.stepIn();
+                        assertNull(reader.next());
+                        reader.stepOut();
+                    }
+                }
+            )
+        );
+    }
+
+    @Test
+    public void symbolTableAnnotationOnNonSymbolTable() throws Exception {
+        readSingleValueOneByteAtATime(
+            toBytes("$ion_symbol_table::[123]"),
+            new IonReaderAssertion() {
+                @Override
+                public void evaluate(IonReader reader) {
+                    assertEquals(IonType.LIST, reader.next());
+                    String[] annotations = reader.getTypeAnnotations();
+                    assertEquals(1, annotations.length);
+                    assertEquals("$ion_symbol_table", annotations[0]);
+                    reader.stepIn();
+                    assertEquals(IonType.INT, reader.next());
+                    assertEquals(123, reader.intValue());
+                    reader.stepOut();
+                }
+            }
+        );
+    }
+
+    @Test
+    public void handlerWithUnlimitedMaxSizeCountsBytesProcessed() throws Exception {
+        AtomicLong totalBytesProcessed = new AtomicLong();
+        maximumBufferSize = Integer.MAX_VALUE;
+        eventHandler = createCountingEventHandler(totalBytesProcessed);
+        byte[] valueBytes = toBytes("[{foo:bar}, (baz zar), 123]");
+        readValuesOneByteAtATime(
+            new Value(
+                valueBytes,
+                new IonReaderAssertion() {
+                    @Override
+                    public void evaluate(IonReader reader) {
+                        assertEquals(IonType.LIST, reader.next());
+                        reader.stepIn();
+                        assertEquals(IonType.STRUCT, reader.next());
+                        reader.stepIn();
+                        assertEquals(IonType.SYMBOL, reader.next());
+                        assertEquals("foo", reader.getFieldName());
+                        assertEquals("bar", reader.stringValue());
+                        reader.stepOut();
+                        assertEquals(IonType.SEXP, reader.next());
+                        assertEquals(IonType.INT, reader.next());
+                        assertEquals(123, reader.intValue());
+                        assertNull(reader.next());
+                        reader.stepOut();
+                    }
+                }
```

Comment:
Given some helper methods, you could write this like this:
```java
                a_list(
                        a_struct(a_field("foo", a_symbol("bar"))),
                        a_sexp(a_symbol("baz"), a_symbol("bar")),
                        an_integer(123)
                )
```

The whole test file written this way reduces it by ~200 lines and is easier to read and write at the expense of slightly increased difficulty of debugging failures. Here's that refactor: https://gist.github.com/jobarr-amzn/7e4fc2fbc91cb31524ff241458849f72

It needs this infrastructure (~200 LOC ;) ) to make an IonReaderAssertion that matches compares an IonValue to an IonReader: https://gist.github.com/jobarr-amzn/26f5f58f71715a34c50e4cc1101b3ab5

Given that this file is not the only place we make structural/content assertions around IonReaders, I think infrastructure like the above could be useful.

Reply:
I appreciate how succinct this proposal makes the tests themselves. However, it goes to great lengths to hide the verbosity of the IonReader API--but that's the API we're testing. We should expect some verbosity in tests that test a verbose API.

I used to try to make my tests as DRY and concise as possible. After having to actually *use* those tests enough times (and with enough time elapsed since writing them), I found that this almost always made them a pain to debug and required more of a time investment to re-familiarize myself with because I had to re-learn what all the helper methods did.

Now, in tests, I always value debuggability above all else. Especially in a code base with as many tests and different authors as ion-java, most developers who encounter a unit test failure will be unfamiliar with that particular test. If such a developer gets a failure in one of these tests, where do they put a breakpoint? Under the proposal, it's not even in the same file; it's in a [recursive assertion helper method](https://gist.github.com/jobarr-amzn/26f5f58f71715a34c50e4cc1101b3ab5#file-ionvaluereaderassertions-java-L43). As the code is currently factored, the assertions are right there in the test and a breakpoint can quickly be placed before any of them without having to do a bunch of digging.

I know there are different philosophies for how to write tests, and I'm not necessarily asserting that mine is the correct one; I'm open to being convinced that it isn't. In the meantime, I'd like to move forward with the current factoring and explore ways to reduce boilerplate without sacrificing debuggability.

Reply:
FWIW I agree about the value of debugging. The proposed recursive helper method definitely makes debugging more difficult. I do think though that this is not a necessary trade-off. More investment in IonValue-based structural assertions could make debugging straightforward while keeping tests easy to read.

I don't expect that anyone should use the IonReader API the way that the tests do- though I'm ready to be wrong on that one. 

> In the meantime, I'd like to move forward with the current factoring and explore ways to reduce boilerplate without sacrificing debuggability.

Sounds good to me. I'll keep this general approach in my pocket and see what I can do to improve it at a later date. Better failure reporting would make all the difference here.

Reply:
> I don't expect that anyone should use the IonReader API the way that the tests do- though I'm ready to be wrong on that one.

I think the tests do some things that a user probably wouldn't often do (e.g. call `next()` multiple times at the end of a container), but generally I think they fairly accurately depict how `IonReader` would be used to traverse values. Maybe I'm not fully understanding this statement. Can you give an example?

Reply:
Sure- in these tests we're making a series of calls to verify arbitrary data, including things like the specific ordering of fields in a struct. That sort of specificity is appropriate when verifying the functioning of a parser, but as a user of ion-java I wouldn't expect to do this. Instead I'd be using higher-level abstractions to guide my use of the library.

That might be ISL, it might be native objects with some other serialization constraints, it might be some other framework, but whatever it is I don't expect to do or see this kind of unrolled/decompiled sequence of calls.